### PR TITLE
APIS-5615 - Decline journey with changes to submission structure

### PR DESCRIPTION
**Pull request by AndySpaven**

### CODE REVIEW

File: app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala [hunk 1]
```diff
@@ -58,13 +65,81 @@ object Submission {
...
+    s.copy(instances = NonEmptyList(delta(s.instances.head), s.instances.tail))
+  }
+
+  val addStatusHistory: (Submission.Status) => Submission => Submission = newStatus => s => {
+    require(Submission.Status.isLegalTransition(s.status, newStatus))
+    changeLatestInstance(_.copy(statusHistory = newStatus :: s.latestInstance.statusHistory))(s)
+  }
+
+  val changeStatusHistory: (Submission.Status => Submission.Status) => Submission => Submission = delta => s => {
+    val inStatus = s.latestInstance.statusHistory.head
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala:96]: same comment as in TPA - should we check this makes a legal state transition?

File: app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala [hunk 2]
```diff
@@ -83,21 +168,43 @@ object Submission {
...
       requestedBy: String
     ) extends Status
+
+    def isLegalTransition(from: Submission.Status, to: Submission.Status): Boolean = (from, to) match {
+      case (c: Created, a: Answering)               => true
+      case (Answering(_, true), s: Submitted)       => true
+      case (s: Submitted, d: Declined)              => true
+      case (s: Submitted, g: Granted)               => true
+      case (s: Submitted, w: GrantedWithWarnings)   => true
+      case (w: GrantedWithWarnings, d: Declined)    => true   // ? Maybe
```
Comment (robdawson-hmrc) [inline app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala:204]: same as TPA - do we know if these are valid?

File: test-utils/uk/gov/hmrc/apiplatform/modules/submissions/AsIdsHelpers.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import uk.gov.hmrc.apiplatform.modules.submissions.domain.models._
+
+trait AsIdsHelpers {
+  implicit class ListQIdSyntax(questionItems: List[QuestionItem]) {
+    def asIds(): List[QuestionId] = {
+      questionItems.map(_.question.id)
+    }
+  }
+
+  implicit class NELQIdSyntax(questionItems: NonEmptyList[QuestionItem]) {
```
Comment (robdawson-hmrc) [inline test-utils/uk/gov/hmrc/apiplatform/modules/submissions/AsIdsHelpers.scala:29]: this is quite a cryptic name, I'm not sure what it represents

File: app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala (hunk 1 above)
Comment (AndySpaven) [reply app/uk/gov/hmrc/apiplatform/modules/submissions/domain/models/Submission.scala:96]: Same response

**Review by robdawson-hmrc — APPROVED** [review]

**Pull request merged**